### PR TITLE
Bindgroup API

### PR DESCRIPTION
**Pull request by mbouron**

### CODE REVIEW

File: libnopegl/src/utils.h [hunk 1]
```diff
@@ -60,6 +60,16 @@
 #define NGLI_MAX(a, b) ((a) > (b) ? (a) : (b))
 #define NGLI_CLAMP(x, min, max) NGLI_MAX(NGLI_MIN(x, max), min)
 
+#define NGLI_ARRAY_MEMDUP(dst, src, name) do {                          \
+    size_t nb = (src).nb_##name;                                        \
```
Comment (ubitux) [inline libnopegl/src/utils.h:64]: `const size_t nb`

File: libnopegl/src/utils.h [hunk 2]
```diff
@@ -60,6 +60,16 @@
...
+    if (nb > 0) {                                                       \
+        (dst).name = ngli_memdup((src).name, nb * sizeof(*(dst).name)); \
+        if (!(dst).name)                                                \
+            return NGL_ERROR_MEMORY;                                    \
+        (dst).nb_##name = nb;                                           \
```
Comment (ubitux) [inline libnopegl/src/utils.h:69]: Assigning the count should probably be done outside the `if` for safety reasons (and maybe assign a `NULL` as well for the pointer)

File: libnopegl/src/utils.h [hunk 3]
```diff
@@ -60,6 +60,16 @@
 #define NGLI_MAX(a, b) ((a) > (b) ? (a) : (b))
 #define NGLI_CLAMP(x, min, max) NGLI_MAX(NGLI_MIN(x, max), min)
 
+#define NGLI_ARRAY_MEMDUP(dst, src, name) do {                          \
```
Comment (ubitux) [inline libnopegl/src/utils.h:63]: Not blocking but I feel like instead we could have the `darray` API with an initializer from a data pointer + count.

`ngli_darray_push_n(struct darray *darray, const void *elements, size_t n);`. You would then have a darrays to work with, which is easier to abstract than the pair of `##name` / `nb_##name` fields

File: libnopegl/src/pipeline_compat.c [hunk 4]
```diff
@@ -39,6 +39,9 @@ struct buffer_binding {
 
 struct pipeline_compat {
     struct gpu_ctx *gpu_ctx;
+    int type;
```
Comment (ubitux) [inline libnopegl/src/pipeline_compat.c:39]: What's kind of type is that? Can you add a comment with the accepted types?

File: libnopegl/src/backends/gl/gpu_ctx_gl.c [hunk 5]
```diff
@@ -995,6 +1001,8 @@ const struct gpu_ctx_class ngli_gpu_ctx_##cls_suffix = {
     .get_preferred_depth_format         = gl_get_preferred_depth_format,         \
     .get_preferred_depth_stencil_format = gl_get_preferred_depth_stencil_format, \
                                                                                  \
+    .set_bindgroup                       = gl_set_bindgroup,                     \
```
Comment (ubitux) [inline libnopegl/src/backends/gl/gpu_ctx_gl.c:1004]: style: extra space

File: libnopegl/src/backends/gl/pipeline_gl.c [hunk 6]
```diff
@@ -22,6 +22,7 @@
 
 #include <string.h>
 
+#include "bindgroup_gl.h"
```
Comment (ubitux) [inline libnopegl/src/backends/gl/pipeline_gl.c:25]: You're moving a lot of code out of `pipeline_gl.c` here; are all the headers still needed?

Same remark in `pipeline_vk.c`

File: libnopegl/src/backends/vk/command_vk.c [hunk 7]
```diff
@@ -42,9 +43,16 @@ void ngli_cmd_vk_freep(struct cmd_vk **sp)
     struct gpu_ctx_vk *gpu_ctx_vk = (struct gpu_ctx_vk *)s->gpu_ctx;
     struct vkcontext *vk = gpu_ctx_vk->vkcontext;
 
+    const struct cb_vk *callbacks = ngli_darray_data(&s->callbacks);
+        for (size_t i = 0; i < ngli_darray_count(&s->callbacks); i++) {
+            const struct cb_vk *callback = &callbacks[i];
+            callback->func(callback->priv);
+    }
```
Comment (ubitux) [inline libnopegl/src/backends/vk/command_vk.c:50]: style: indentation is off

File: libnopegl/src/backends/vk/command_vk.h [hunk 8]
```diff
@@ -26,6 +27,13 @@
 
 #include "darray.h"
 
+typedef void (*vk_cb_func)(void *priv);
+
+struct cb_vk {
+  vk_cb_func func;
+  void *priv;
+};
```
Comment (ubitux) [inline libnopegl/src/backends/vk/command_vk.h:35]: What's going on with your indent? 😅

File: libnopegl/src/backends/vk/bindgroup_vk.h [hunk 9]
```diff
@@ -39,7 +40,7 @@ struct bindgroup_layout_vk {
 
 struct bindgroup_vk {
     struct bindgroup parent;
-    int64_t refcount;
+    size_t refcount;
```
Comment (ubitux) [inline libnopegl/src/backends/vk/bindgroup_vk.h]: do you mind splitting this out?

File: libnopegl/src/backends/vk/buffer_vk.c [hunk 10]
```diff
@@ -229,19 +230,34 @@ void ngli_buffer_vk_unmap(struct buffer *s)
     vkUnmapMemory(vk->device, s_priv->memory);
 }
 
+void ngli_buffer_vk_ref(struct buffer *s)
+{
+    struct buffer_vk *s_priv = (struct buffer_vk *)s;
+    s_priv->refcount++;
+}
+
+void ngli_buffer_vk_unref(struct buffer *s)
```
Comment (ubitux) [inline libnopegl/src/backends/vk/buffer_vk.c:239]: Without an `unrefp` you will have the issue that the pointer will always remain: the `ngli_freep()` below will only clear the local pointer.

**Review by ubitux — CHANGES_REQUESTED** [review]

File: libnopegl/src/pipeline_compat.c [hunk 11]
```diff
@@ -129,6 +143,30 @@ int ngli_pipeline_compat_init(struct pipeline_compat *s, const struct pipeline_c
         s->nb_vertex_buffers = nb_attributes;
     }
 
+    const size_t nb_buffers = pipeline_resources->nb_buffers;
+    if (nb_buffers) {
+        s->buffers = ngli_calloc(nb_buffers, sizeof(struct buffer_binding));
```
[the remark this answers is not in the document]
Comment (mbouron) [reply libnopegl/src/pipeline_compat.c]: Done.

File: libnopegl/src/backends/vk/pipeline_vk.c [hunk 12]
```diff
@@ -864,7 +864,7 @@ static int prepare_and_bind_descriptor_set(struct pipeline *s, VkCommandBuffer c
     return 0;
 }
 
-static int prepare_pipeline(struct pipeline *s, VkCommandBuffer cmd_buf)
+static int prepare_and_bind_graphcs_pipeline(struct pipeline *s, VkCommandBuffer cmd_buf)
```
[the remark this answers is not in the document]
Comment (mbouron) [reply libnopegl/src/backends/vk/pipeline_vk.c]: Fixed.

File: libnopegl/src/pipeline_compat.c [hunk 13]
```diff
@@ -129,6 +143,30 @@ int ngli_pipeline_compat_init(struct pipeline_compat *s, const struct pipeline_c
...
+        for (size_t i = 0; i < nb_buffers; i++) {
+            struct buffer *buffer = pipeline_resources->buffers[i];
```
[the remark this answers is not in the document]
Comment (mbouron) [reply libnopegl/src/pipeline_compat.c]: Done.

File: libnopegl/src/utils.h (hunk 1 above)
Comment (mbouron) [reply libnopegl/src/utils.h]: Done.

File: libnopegl/src/backends/gl/gpu_ctx_gl.c (hunk 5 above)
Comment (mbouron) [reply libnopegl/src/backends/gl/gpu_ctx_gl.c]: Done.

File: libnopegl/src/backends/vk/command_vk.c (hunk 7 above)
Comment (mbouron) [reply libnopegl/src/backends/vk/command_vk.c]: Done.

File: libnopegl/src/backends/vk/command_vk.h (hunk 8 above)
Comment (mbouron) [reply libnopegl/src/backends/vk/command_vk.h]: Done.

File: libnopegl/src/backends/vk/bindgroup_vk.h (hunk 9 above)
Comment (mbouron) [reply libnopegl/src/backends/vk/bindgroup_vk.h]: Fixed.

File: libnopegl/src/utils.h (hunk 2 above)
Comment (mbouron) [reply libnopegl/src/utils.h:71]: I assigned unconditionally the ptr to NULL and nb to 0 at the beginning.

File: libnopegl/src/backends/vk/buffer_vk.c (hunk 10 above)
Comment (mbouron) [reply libnopegl/src/backends/vk/buffer_vk.c]: Dropped.

File: libnopegl/src/distmap.c [hunk 14]
```diff
@@ -640,7 +640,7 @@ int ngli_distmap_finalize(struct distmap *s)
     if (!s->pipeline_compat)
         return NGL_ERROR_MEMORY;
 
-    const struct pipeline_resources pipeline_resources = ngli_pgcraft_get_pipeline_resources(s->crafter);
+    const struct pipeline_compat_resources pipeline_resources = ngli_pgcraft_get_pipeline_resources(s->crafter);
```
[the remark this answers is not in the document]
Comment (mbouron) [reply libnopegl/src/distmap.c:643]: I originally planned to rename them to `ngli_pgcraft_get_pipeline_compat_layout()` and `ngli_pgcraft_get_pipeline_compat_resources` but I kept the current name as it does not bring too much to the table. I plan to rename the pipeline_compat structure to something else as well as `pgcraft_compat_info` (if i'm not removing it).

File: libnopegl/src/utils.h (hunk 3 above)
Comment (mbouron) [reply libnopegl/src/utils.h:63]: Indeed, a darray_dup() or darray_copy() is easier to work with but it only works with arrays on the heap. An alternative would be something like:
``` c
#define NGLI_ARRAY(type) struct { type *elems; size_t nb_elems; } and have a bunch of macro associated with it (like `NGLI_ARRAY_MEMDUP()`, `NGLI_ARRAY_APPEND()`, ...)

This has the benefit of keeping the type and you can style do something like this:
```
NGLI_ARRAY(int) array = {.elems = (int[]){1, 2}, .nb_elems = 2};
```
If you don't want to deal with alloc/dealloc.

**Review by ubitux — APPROVED** [review]

**Pull request merged**